README.md
# grunt-appdmg
> Grunt plugin for generating Mac OS X DMG-images

[![NPM Version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Dependency Status][deps-image]][deps-url]

## Overview
[node-appdmg](https://github.com/LinusU/node-appdmg) is an awesome command line tool to generate Mac disk images.
This Grunt plugin wraps the node-appdmg and executes it using Gruntfile.

You can use [Grunt template strings](http://gruntjs.com/api/grunt.template) in the appdmg config, like: `title: '<%= pkg.name %>'`.

**Note:**  
Currently grunt-appdmg works on **Mac OS X only** due to the [node-appdmg limitation](https://github.com/LinusU/node-appdmg/issues/14).

## Getting Started

This plugin requires Grunt `>=0.4.0`.  
If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide.

### Installation
```shell
$ npm install grunt-appdmg --save-dev
```

If you want to include this plugin in a cross platform project, install it with `--save-optional` instead of `--save-dev`.
This will prevent installation error on Windows/Linux.

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:
```js
grunt.loadNpmTasks('grunt-appdmg');
```

## The "appdmg" task

### Options
See the **[JSON Specification](https://github.com/LinusU/node-appdmg#json-specification)** of node-appdmg.  
Additionally `basepath` option is available.

#### basepath
Type: `String`  
Default: `process.cwd()` - directory that contains Gruntfile.js

Optional. Base path to look for asset files: `icon`, `background` and `contents.path`.

### Example config
In your project's Gruntfile, add a section named `appdmg` to the data object passed into `grunt.initConfig()`.
```js
grunt.initConfig({
  appdmg: {
    options: {
      basepath: 'path/to/assets',
      title: 'Title of DMG',
      icon: 'icon.icns',
      background: 'background.png',
      'icon-size': 80,
      contents: [
        {x: 448, y: 344, type: 'link', path: '/Applications'},
        {x: 192, y: 344, type: 'file', path: 'your-app.app'},
        {x: 512, y: 128, type: 'file', path: 'extra-file.txt'}
      ]
    },
    target: {
      dest: 'path/to/your-app.dmg'
    }
  }
});
```

## License
Copyright (c) 2014-2015 Rakuten, Inc. Licensed under the [MIT License](LICENSE).

[npm-image]: https://img.shields.io/npm/v/grunt-appdmg.svg?style=flat
[npm-url]: https://www.npmjs.com/package/grunt-appdmg
[travis-image]: https://img.shields.io/travis/rakuten-frontend/grunt-appdmg/master.svg?style=flat
[travis-url]: https://travis-ci.org/rakuten-frontend/grunt-appdmg
[deps-image]: http://img.shields.io/david/rakuten-frontend/grunt-appdmg.svg?style=flat
[deps-url]: https://david-dm.org/rakuten-frontend/grunt-appdmg


CHANGELOG
v0.3.1:
  date: 2015-03-24
  changes:
    - Specify the supported OS in package.json.
v0.3.0:
  date: 2015-02-09
  changes:
    - Bump node-appdmg to v0.3.0.
    - Add `basepath` option.
    - Deprecate `configFile` option and avoid creating temporary json file.
v0.2.0:
  date: 2014-10-09
  changes:
    - Release as open source.
    - Upgrade node-appdmg to v0.2.2.
v0.1.0:
  date: 2014-06-19
  changes:
    - Initial release.


test/appdmg_test.js
'use strict';

var fs = require('fs');
var grunt = require('grunt');

/*
  ======== A Handy Little Nodeunit Reference ========
  https://github.com/caolan/nodeunit

  Test methods:
    test.expect(numAssertions)
    test.done()
  Test assertions:
    test.ok(value, [message])
    test.equal(actual, expected, [message])
    test.notEqual(actual, expected, [message])
    test.deepEqual(actual, expected, [message])
    test.notDeepEqual(actual, expected, [message])
    test.strictEqual(actual, expected, [message])
    test.notStrictEqual(actual, expected, [message])
    test.throws(block, [error], [message])
    test.doesNotThrow(block, [error], [message])
    test.ifError(value)
*/

exports.appdmg = {

  setUp: function (done) {
    // setup here if necessary
    done();
  },

  basic: function (test) {
    test.expect(1);
    var exists = fs.existsSync('test/tmp/basic.dmg');
    test.ok(exists, 'should create a dmg file with basic option.');
    test.done();
  },

  basepath: function (test) {
    test.expect(1);
    var exists = fs.existsSync('test/tmp/basepath.dmg');
    test.ok(exists, 'should create a dmg file with basepath option.');
    test.done();
  },

  extra: function (test) {
    test.expect(1);
    var exists = fs.existsSync('test/tmp/extra.dmg');
    test.ok(exists, 'should create a dmg file including extra contents.');
    test.done();
  }

};


tasks/appdmg.js
'use strict';

var path = require('path');
var appdmg = require('appdmg');
var chalk = require('chalk');

module.exports = function (grunt) {

  // Please see the Grunt documentation for more information regarding task
  // creation: http://gruntjs.com/creating-tasks
  grunt.registerMultiTask('appdmg', 'Generate DMG-images for Mac OSX', function () {

    var options = this.options();
    var done = this.async();

    if (options.configFile) {
      grunt.log.warn('"configFile" option has been deprecated.');
      delete options.configFile;
    }

    var basepath = options.basepath || process.cwd();
    delete options.basepath;

    // Iterate over all specified file groups.
    this.files.forEach(function (filePair) {

      var dirname = path.dirname(filePair.dest);
      var emitter;

      // Create directory beforehand to prevent error.
      grunt.file.mkdir(dirname);

      // Run appdmg module.
      // TODO: Remove logging and use native appdmg's method in the future release.
      emitter = appdmg({basepath: basepath, specification: options, target: filePair.dest});
      emitter.on('progress', function (info) {
        if (info.type === 'step-begin') {
          var line =  '[' + (info.current <= 9 ? ' ' : '') + info.current + '/' + info.total + '] ' + info.title + '...';
          grunt.log.write(line + String.repeat(' ', 45 - line.length));
        }
        if (info.type === 'step-end') {
          var op = ({
            ok: ['green', ' OK '],
            skip: ['yellow', 'SKIP'],
            error: ['red', 'FAIL']
          }[info.status]);
          grunt.log.write('[' + chalk[op[0]](op[1]) + ']\n');
        }
      });
      emitter.on('finish', function (image) {
        grunt.log.writeln('\nImage: ' + chalk.cyan(filePair.dest) + ' was created');
        done();
      });
      emitter.on('error', function (info) {
        grunt.log.error('Error');
        done(false);
      });

    });

  });

};


Gruntfile.js
'use strict';

module.exports = function (grunt) {

  // load all npm grunt tasks
  require('load-grunt-tasks')(grunt);

  // Project configuration.
  grunt.initConfig({

    pkg: grunt.file.readJSON('package.json'),

    // JSHint
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>'
      ],
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      }
    },

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['test/tmp']
    },

    // Configuration to be run (and then tested).
    appdmg: {
      basic: {
        options: {
          title: '<%= pkg.name %> Test',
          background: 'test/fixtures/TestBkg.png',
          'icon-size': 80,
          contents: [
            {
              x: 192,
              y: 344,
              type: 'file',
              path: 'test/fixtures/TestApp.app'
            },
            {
              x: 448,
              y: 344,
              type: 'link',
              path: '/Applications'
            }
          ]
        },
        dest: 'test/tmp/basic.dmg'
      },
      basepath: {
        options: {
          basepath: 'test/fixtures/',
          title: '<%= pkg.name %> Test',
          background: 'TestBkg.png',
          'icon-size': 80,
          contents: [
            {
              x: 192,
              y: 344,
              type: 'file',
              path: 'TestApp.app'
            },
            {
              x: 448,
              y: 344,
              type: 'link',
              path: '/Applications'
            }
          ]
        },
        dest: 'test/tmp/basepath.dmg'
      },
      extra: {
        options: {
          basepath: 'test/fixtures/',
          title: '<%= pkg.name %> Test',
          icon: 'TestIcon.icns',
          background: 'TestBkg.png',
          'icon-size': 80,
          contents: [
            {
              x: 192,
              y: 344,
              type: 'file',
              path: 'TestApp.app'
            },
            {
              x: 448,
              y: 344,
              type: 'link',
              path: '/Applications'
            },
            {
              x: 512,
              y: 128,
              type: 'file',
              path: 'TestDoc.txt'
            }
          ]
        },
        dest: 'test/tmp/extra.dmg'
      }
    },

    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js']
    }

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // Whenever the "test" task is run, first clean the "test/tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['jshint', 'clean', 'appdmg', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['test']);

};
